### PR TITLE
`Programming exercises`: Limit Jenkins builds to default branches

### LINKED ISSUE #4088: Conflicting Branches & Build Plans
### Describe the bug

One student has pushed their solution to another branch than the configured `master` branch (namely the `main` branch (I assume he/she has done this because that's the new default branch name)). Since the system is configured to use the `master` branch, I've encountered a strange state:

image

Obviously, the tests have been executed but on the main branch. But Artemis does not show the solution (Since `master` and `main` branch exists in the repository)

### To Reproduce

See description.

### Expected behavior

Artemis shall show the code of the rating or no test results. But a consistent state :) 

### Screenshots

_No response_

### What browsers are you seeing the problem on?

Other (specify in "Additional context")

### Additional context

Artemis 5.1.5

### Relevant log output

_No response_

### LINKED ISSUE #5575: `Programming Exercises`: Submissions for commits on all branches are created
### Describe the bug

Artemis does not only create submissions for commits on the main/default branch, but also for commits on other branches.

This is a high priority bug, as students that commit to other branches than the default see a positive result in Artemis. However, when a tutor manually assesses the exercise, they only see the default branch in the online editor where nothing has been changed.
This could also happen in exams where students are allowed to use the local IDE to solve the task.

### To Reproduce

1. Create a programming exercise.
2. As student, commit to a non-default branch in the student repo.
3. Artemis creates a submission for this commit.

### Expected behavior

Artemis only creates new submissions for commits on the default branch and ignores all other ones.

### Screenshots

_No response_

### What browsers are you seeing the problem on?

Firefox

### Additional context

- Artemis 5.9.8 (while investigating the bug I noticed that it seems to have occurred at least once, a month before as well, so all versions at least since 5.9.0 probably have the same issue)
- Gitlab/Jenkins setup

### Relevant log output

```shell
08:00:36.303 INFO t.i.w.a.s.p.ProgrammingSubmissionService : NotifyPush invoked due to the commit 457652989a86cfee17eb3ee8d306ded00dfa3d8b by Benedikt Fein with <...> in branch testbranch
08:00:36.304 WARN .i.w.a.w.r.ProgrammingSubmissionResource : Processing submission for participation 1469 failed: Submission for participation id 1469 in branch testbranch will be ignored! Only the default branch is considered
08:00:57.547 INFO d.t.i.w.artemis.web.rest.ResultResource  : Artemis received a new result for build plan TEST01GGTJAVA17-FEIN02
08:01:05.093 WARN .a.s.p.ProgrammingExerciseGradingService : Could not find pending ProgrammingSubmission for Commit Hash Optional[457652989a86cfee17eb3ee8d306ded00dfa3d8b] (Participation 1469, Build Plan TEST01GGTJAVA17-FEIN02). Will create a new one subsequently...
08:01:06.338 INFO d.t.i.w.artemis.web.rest.ResultResource  : The new result for TEST01GGTJAVA17-FEIN02 was saved successfully
```

### PR DESCRIPTION
**Pull request by krusche**
Close #5575 and close #4088

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming exercise

1. As a student: change the branch to a random other one (NOT the default one) and commit and push your changes
2. Jenkins or Bamboo should NOT build the changes
3. Artemis should NOT have a submission and should NOT have a result for the submission on the wrong branch

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java [hunk 1]
```diff
@@ -44,6 +43,8 @@ public class ModelFactory {
 
     public static final String USER_PASSWORD = "00000000";
 
+    public static final String defaultBranch = "main";
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java:46]:
```suggestion
    public static final String DEFAULT_BRANCH = "main";
```

File: src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java [hunk 2]
```diff
@@ -997,62 +998,38 @@ public static LockRepositoryPolicy generateLockRepositoryPolicy(int submissionLi
...
+        final var testSuite = new TestSuiteDTO("TestSuiteName1", now().toEpochSecond(), 0, 0, failedTestNames.size(), successfulTestNames.size() + failedTestNames.size(),
+                new ArrayList<>());
+        testSuite.testCases().addAll(successfulTestNames.stream().map(name -> new TestCaseDTO(name, "Class", 0d)).toList());
+        testSuite.testCases().addAll(failedTestNames.stream()
+                .map(name -> new TestCaseDTO(name, "Class", 0d, new ArrayList<>(), List.of(new TestCaseDetailMessageDTO(name + " error message")), new ArrayList<>())).toList());
+
+        final var commitDTO = new CommitDTO(TestConstants.COMMIT_HASH_STRING, repoName, defaultBranch);
+        final var staticCodeAnalysisReports = enableStaticAnalysisReports ? generateStaticCodeAnalysisReports(programmingLanguage) : new ArrayList<StaticCodeAnalysisReportDTO>();
+
+        return new TestResultsDTO(successfulTestNames.size(), 0, 0, failedTestNames.size(), fullName, commits != null && commits.size() > 0 ? commits : List.of(commitDTO),
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java:1025]:
```suggestion
        return new TestResultsDTO(successfulTestNames.size(), 0, 0, failedTestNames.size(), fullName, commits == null || commits.isEmpty() ? List.of(commitDTO) : commits,
```

File: src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java [hunk 3]
```diff
@@ -997,62 +998,38 @@ public static LockRepositoryPolicy generateLockRepositoryPolicy(int submissionLi
     /**
      * Creates a dummy DTO used by Jenkins, which notifies about new programming exercise results.
      *
-     * @param repoName name of the repository
-     * @param successfulTestNames names of successful tests
-     * @param failedTestNames names of failed tests
-     * @param programmingLanguage programming language to use
+     * @param repoName                    name of the repository
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java:1001]:
```suggestion
     * @param fullName                    ???
     * @param repoName                    name of the repository
```
The `@param` is missing, but I‘m not entirely sure what that parameter is.

File: src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java [hunk 4]
```diff
@@ -1069,166 +1046,88 @@ public static TestResultsDTO generateTestResultDTO(String repoName, List<String>
...
-        notification.setNotificationType("TestNotification");
-        notification.setBuild(build);
-        notification.setPlan(plan);
+        final var job = new BambooBuildResultNotificationDTO.BambooJobDTO(42, failedTests, successfulTests, new ArrayList<>(), new ArrayList<>(), new ArrayList<>());
+        final var vcs = new BambooBuildResultNotificationDTO.BambooVCSDTO(TestConstants.COMMIT_HASH_STRING, repoName, defaultBranch, new ArrayList<>());
+        final var plan = new BambooBuildPlanDTO(planKey != null ? planKey : "TEST201904BPROGRAMMINGEXERCISE6-STUDENT1");
 
-        return notification;
+        final var build = new BambooBuildResultNotificationDTO.BambooBuildDTO(false, 42, "foobar", buildCompletionDate != null ? buildCompletionDate : now().minusSeconds(5),
+                failedTestNames.isEmpty(), summary, vcsDtos != null && vcsDtos.size() > 0 ? vcsDtos : List.of(vcs), List.of(job));
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java:1097]:
```suggestion
                failedTestNames.isEmpty(), summary, vcsDtos == null || vcsDtos.isEmpty() ? List.of(vcs) : vcsDtos, List.of(job));
```

**Review by b-fein — APPROVED** [review]
Jenkins builds are now only triggered for the default branch. The match is exact, i.e., committing to branches that contain ‘main’ somewhere in their name does not trigger a commit either.

I did not code-review the changes to Bamboo-DTOs but only changes to other classes/files.

**Review by sleiss — APPROVED** [review]
Code changes look good to me.

**Pull request merged**